crates/simulate/src/state/mod.rs

```diff
@@ -1,249 +1,367 @@
+    pub fn timestamp(&self) -> U256 {
         U256::from(self.evm.env.block.timestamp)
     }
 
-    pub fn read_block_number(&self) -> U256 {
+    pub fn block_number(&self) -> U256 {
         U256::from(self.evm.env.block.number)
     }
 
-    pub fn simulate_evm_tx_ref(
-        &self,
-        tx: &TxEnv,
-        env: Option<Env>,
-    ) -> SimStateResult<ExecutionResult> {
-        // Create a sim_evm with no db and cloned and/or passed env (fairly cheap).
-        let env = env.unwrap_or(self.evm.env.clone());
-        let mut sim_evm = EVM::with_env(env);
-        // Set sim_evm's db to a ref of the actual evm's db.
-        sim_evm.database(self.get_read_db()?);
-        // Update env to new tx.
-        sim_evm.env.tx = tx.clone();
-        // We can now use the sim_evm to simulate the tx without writing to db.
-        match sim_evm.transact_ref() {
-            Ok(result_and_state) => Ok(result_and_state.result),
-            Err(e) => Err(SimStateError::EvmStateError(anyhow::format_err!(
-                "Error while simulating evm tx: {:?}",
-                e
-            ))),
-        }
+    /// Update the evm's block number and timestamp.
+    pub fn update_time(&mut self, time_env: TimeEnv) {
+        self.evm.env.block.number = time_env.block_number.into();
+        self.evm.env.block.timestamp = time_env.block_timestamp.into();
     }
 
-    /// Update the time env.
-    /// # Arguments
-    /// * `time_env` - The time env.
-    pub fn update_time_env(&mut self, time_env: TimeEnv) {
-        self.evm.env.block.number = time_env.number.into();
-        self.evm.env.block.timestamp = time_env.timestamp.into();
-    }
-
-    // Add an account to evm.
-    pub fn insert_account_info(
-        &mut self,
-        address: Address,
-        account_info: AccountInfo,
-    ) -> SimStateResult<()> {
+    /// Add an account to the evm.
+    pub fn add_account(&mut self, address: Address, account_info: AccountInfo) -> StateResult<()> {
         self.evm
             .db()
-            .ok_or(SimStateError::EvmDBNotSet)?
+            .ok_or(StateError::EvmDBNotSet)?
             .insert_account_info(address.into(), account_info);
 
         Ok(())
     }
 
-    /// Execute a transaction in the execution environment.
-    /// # Arguments
-    /// * `tx` - The transaction environment that is used to execute the transaction.
-    /// # Returns
-    /// * `ExecutionResult` - The execution result of the transaction.
-    pub fn execute_evm_tx(&mut self, tx: TxEnv) -> SimStateResult<ExecutionResult> {
+    pub fn call_evm_tx_raw(&self, tx: TxEnv) -> StateResult<EvmTxOutput> {
+        let mut sim_evm = EVM::with_env(self.evm.env.clone());
+        sim_evm.database(self.get_read_db()?);
+        sim_evm.env.tx = tx;
+
+        let output = sim_evm
+            .transact_ref()
+            .map_err(StateError::EvmExecutionError)?;
+        unpack_evm_tx_output(output.result)
+    }
+
+    pub fn call_evm_tx<D: Detokenize>(
```

Comment:
This is equivalent to what was previously `simulate_evm_tx_ref`, but allows the agents to use an ethers `ContractCall`. The `call_evm_tx_and_decode` takes the additional step of decoding the output assuming the call is a success.